com.unity.formats.fbx/Tests/FbxTests/ModelExporterTest.cs

```diff
@@ -1227,5 +1228,72 @@ public void TestMissingMeshExport()
             ModelExporter.ExportObject(filename, cube);
             Assert.IsNotNull(filename);
         }
+
+        [Test]
+        public void TestMaterialScaleAndOffset() 
+        {
+            string matClampGuid = "e6598d8bd228e5940988877e9eddd594";
+            string matRepeatGuid = "f07bd71d1d87a7b41acfc483bf06be3f";
+            
+            var filename = GetRandomFbxFilePath();
+
+            var clampMat = AssetDatabase.LoadAssetAtPath<Material>(AssetDatabase.GUIDToAssetPath(matClampGuid));
+            var cubeClamp = GameObject.CreatePrimitive(PrimitiveType.Cube);
+            cubeClamp.name = "Clamp";
+            cubeClamp.GetComponent<MeshRenderer>().sharedMaterial = clampMat;
+
+            var repeatMat = AssetDatabase.LoadAssetAtPath<Material>(AssetDatabase.GUIDToAssetPath(matClampGuid));
+            var cubeRepeat = GameObject.CreatePrimitive(PrimitiveType.Cube);
+            cubeRepeat.name = "Repeat";
+            cubeRepeat.GetComponent<MeshRenderer>().sharedMaterial = repeatMat;
+
+            var settings = new ExportModelSettingsSerialize();
+            settings.SetEmbedTextures(true);
+            ModelExporter.ExportObjects(filename, new Object[] { cubeClamp, cubeRepeat }, settings);
+
+            var modelImporter = AssetImporter.GetAtPath(filename) as ModelImporter;
+            modelImporter.ExtractTextures(tempUnpackPath);
+            AssetDatabase.Refresh();
+            
+            GameObject fbxObj = AssetDatabase.LoadMainAssetAtPath(filename) as GameObject;
+            var importedClampMat = fbxObj.transform.Find("Clamp").GetComponent<MeshRenderer>().sharedMaterial;
+            var importedRepeatMat = fbxObj.transform.Find("Repeat").GetComponent<MeshRenderer>().sharedMaterial;
+
+            // Ensure we're actually checking on the newly imported textures, not on the ones used for export
+            Assert.NotNull(importedClampMat.mainTexture);
+            Assert.NotNull(importedRepeatMat.mainTexture);
+            
+            // (Case 1424290) This fails - seems there's a bug with modelImporter.ExtractTextures when used inside a test
+            // Manually exporting the FBX and extracting textures via Inspector properly produces separate textures.
+            // Assert.AreNotSame(importedClampMat.mainTexture, importedRepeatMat.mainTexture);
+            
+            // Seems these also randomly fail? Sometimes work, sometimes not...
+            // Assert.AreNotSame(clampMat.mainTexture, importedClampMat.mainTexture);
```

Comment:
the texture extraction in the test seems like it might not update the material to point to the extracted textures. If I run the test without deleting the generated asset, the materials seem to point to the original, not the embedded ones. If I export and extract the textures manually, it points to the extracted textures.

Reply:
oh sorry this is the bug commented right above, I misunderstood. I'm not sure if waiting a few frames would solve the issue, I think there is an extra remap done after ExtractTextures is called when you click the button in the UI.

Extract textures call: https://github.com/Unity-Technologies/UnityCsReference/blob/master/Modules/AssetPipelineEditor/ImportSettings/ModelImporterMaterialEditor.cs#L299

Remap call: https://github.com/Unity-Technologies/UnityCsReference/blob/master/Modules/AssetPipelineEditor/ImportSettings/ModelImporterMaterialEditor.cs#L326

I'm not sure if the remap is easy to replicate here, or maybe could find the matching textures in the folder they were extracted to and load the texture asset directly to check the values. If it is difficult to do, then in my opinion since there is the other bug with the wrap mode not importing correctly, this could also potentially be fixed when that issue is resolved.

Reply:
Turns out it was easy to replicate - thanks for pointing me at the source there, wasn't aware that Extract exactly only does that and not the remapping part. One bug down! Only the import wrap mode issue remains, but that code can just be uncommented once that's at some point fixed in core.